Ghidra/Processors/WebAssembly/data/languages/WebAssembly.sinc

```diff
@@ -0,0 +1,426 @@
+	export *[register]:8 tmp_sp;
+}
+
+br_target: ctx_br_target is ctx_br_target {
+	export *[ram]:4 ctx_br_target;
+}
+
+funcidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+typeidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+elemidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+tableidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+tableidx2: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+dataidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+memidx: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+memidx2: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+count: ULeb128 is ULeb128 {
+	export ULeb128;
+}
+
+reftype: "reftype="^bytev is bytev {
+	export *[const]:1 bytev;
+}
+
+# An explicit no-op operation, to be used before
+# callother instructions to work around Ghidra issue #3390
+# (which causes incorrect decompilation if the first pcodeop
+# in a branch target instruction is a callother)
+macro nop() {
+	LR = LR;
+}
+
+macro trap() {
+<trap_loop>
+	halt_trap();
+	goto <trap_loop>;
+}
+
+macro doBranch(br_target) {
+	nop();
+	popCallOther(&t0);
+	pushCallOther(&t0);
+	goto br_target;
+}
+
+################################################################
+
+## Locals directives at the start of the function
+:.locals is count; ctx_is_directive=1; ctx_directive_type=$(DIRECTIVE_LOCALS) {
+}
+
+:.local "count="^count "type="^typeidx is count; typeidx; ctx_is_directive=1; ctx_directive_type=$(DIRECTIVE_LOCAL) {
+}
+
+## Control instructions
+:unreachable is opc=0x00; ctx_is_directive=0 {
+	trap();
+}
+
+:nop is opc=0x01; ctx_is_directive=0 {
+}
+
+:block "blocktype="^ULeb128 is opc=0x02; ULeb128; ctx_is_directive=0 {
+}
+:loop "blocktype="^ULeb128 is opc=0x03; ULeb128; ctx_is_directive=0 {
+}
+:if "blocktype="^ULeb128 is opc=0x04; ULeb128; br_target; sp1_32; ctx_is_directive=0 {
+	if(sp1_32 != 0) goto inst_next;
+	doBranch(br_target);
+}
+:else is opc=0x05; br_target; ctx_is_directive=0 {
+	doBranch(br_target);
+}
+
+:end is opc=0x0b; ctx_is_return=1; ctx_is_directive=0 {
+	nop();
+	popCallOther(&o0);
+	return [LR];
+}
+:end is opc=0x0b; ctx_is_return=0; ctx_is_directive=0 {
+}
+
+:br br_target is opc=0x0c; ULeb128; br_target; ctx_is_directive=0 {
+	doBranch(br_target);
+}
+:br_if br_target is opc=0x0d; ULeb128; br_target; sp1_32; ctx_is_directive=0 {
+	if(sp1_32 == 0) goto inst_next;
+	doBranch(br_target);
+}
+
+:.case ctx_case_index br_target is ULeb128; br_target; ctx_case_index; ctx_is_directive=1; ctx_directive_type=$(DIRECTIVE_CASE) {
+	if(BRTableInput != ctx_case_index) goto inst_next;
+	doBranch(br_target);
+}
+
+:.default br_target is ULeb128; br_target; ctx_is_directive=1; ctx_directive_type=$(DIRECTIVE_DEFAULT) {
+	doBranch(br_target);
+}
+
+:br_table is opc=0x0e; ULeb128; sp1_32; ctx_is_directive=0 {
+	BRTableInput = sp1_32;
+}
+
+:return is opc=0x0f; ctx_is_directive=0 {
+	nop();
+	popCallOther(&o0);
+	return [LR];
+}
+:call br_target is opc=0x10; ULeb128; br_target; ctx_is_directive=0 {
+	nop();
+	popCallOther(&i0);
+	call br_target;
```

Comment:
`LR` is a totally artificial register, so setting LR is actually not necessary because Ghidra will analyze `call`s using the appropriate control flow regardless of any LR register. It's set for most processor modules because it's a real register whose value might matter; in Wasm, the LR register is entirely unobservable so setting it doesn't have any use.

I'm happy to be proven wrong though - I imagine that not setting `LR` might break things like inlined functions, but those are broken anyway so I'm not sure it will matter.

Reply:
emulator usage will need this

Reply:
Fixed via extra emulator support code.